In [1]:
from fractions import Fraction
import numpy as np
import sympy as sym
from scipy.spatial.distance import euclidean
from sympy import I
from sympy.ntheory import factorint

**Содержание**<a id='toc0_'></a>    
- [__Линейные отображения__](#toc1_1_)    
  - [__Линейные отображения на плоскости__](#toc1_1_1_)    
- [__Умножение матриц__](#toc1_2_)    
- [__Базис, размерность, ранг__](#toc1_3_)    
- [__Определитель и его свойства__](#toc1_4_)    
- [__Переход в другой базис__](#toc1_5_)    
  - [__Переход от произвольного базиса к стандартному__](#toc1_5_1_)    
  - [__Переход от стандартного базиса к произвольному__](#toc1_5_2_)    
  - [__Переход от произвольного базиса к другому произвольному__](#toc1_5_3_)    
- [__Длина, углы, скалярное произведение__](#toc1_6_)    
  - [__Теорема косинусов__](#toc1_6_1_)    
- [__Ортогональные преобразования__](#toc1_7_)    
- [__Комплексные числа__](#toc1_8_)    

<!-- vscode-jupyter-toc-config
	numbering=false
	anchor=true
	flat=false
	minLevel=1
	maxLevel=6
	/vscode-jupyter-toc-config -->
<!-- THIS CELL WILL BE REPLACED ON TOC UPDATE. DO NOT WRITE YOUR TEXT IN THIS CELL -->

***
## <a id='toc1_1_'></a>[__Линейные отображения__](#toc0_)
_Линейным отображением_ векторного пространства $V$ в векторное пространство $W$ называется __функция__ $f : V \rightarrow W$, удовлетворяющая следующим двум условиям:

1. $f(\vec{x}) + f(\vec{y}) = f(\vec{x} + \vec{y})$ для любых $\vec{x}, \vec{y} \in V$ ($f$ _respects addition_),

2. $f(c\vec{x}) = cf(\vec{x})$ для любого $\vec{x} \in V$ и любого числа $c \in \R$ ($f$ _respects multiplication by scalar_).

Заметим, что в левой части равенства _1_ сложение векторов происходит в векторном пространстве $W$, а в правой части сложение происходит в $V$.

Любое л.о. $f : \R \rightarrow \R$ задается одним числом.

Любое л.о. $f : \R \rightarrow \R^m$ задается столбцом высоты $m$, состоящим из чисел.

Любое л.о. $f : \R^n \rightarrow \R$ задается строкой длины $n$, состоящей из чисел.

Любое л.о. $f : \R^n \rightarrow \R^m$ задается таблицей (матрицей) с $m$ строками и $n$ столбцами, заполненной числами.

Любая матрица задаёт линейное отображение. Любое линейное отображение задаётся какой-то матрицей. Разные матрицы задают разные линейные отображения. Есть взаимно  однозначное соответствие между матрицами и линейными отображениями: каждой матрице соответствует ровно одно линейное отображение, и каждому линейному отображению поставлена в соответствие ровно одна матрица.

$f : \R^2 \rightarrow \R^2$ по формуле $f(x_1,x_2) = (x_2,x_1)$:

In [2]:
x = sym.Matrix([sym.Indexed('x', i) for i in (1, 2)])
x

Matrix([
[x[1]],
[x[2]]])

In [3]:
f = sym.Matrix([[0, 1], [1, 0]])
f

Matrix([
[0, 1],
[1, 0]])

In [4]:
f * x

Matrix([
[x[2]],
[x[1]]])

$f : \R^3 \rightarrow \R^2$, где $f(x_1,x_2,x_3) = (x_1 - x_2 + 2x_3, 10x_2)$:

In [5]:
x = sym.Matrix([sym.Indexed('x', i) for i in range(1, 4)])
x

Matrix([
[x[1]],
[x[2]],
[x[3]]])

In [6]:
f = sym.Matrix([[1, -1, 2], [0, 10, 0]])
f

Matrix([
[1, -1, 2],
[0, 10, 0]])

In [7]:
f @ x

Matrix([
[x[1] - x[2] + 2*x[3]],
[             10*x[2]]])

***
### <a id='toc1_1_1_'></a>[__Линейные отображения на плоскости__](#toc0_)

$f(\vec{x}) = k\vec{x}$: $f$ растягивает все векторы в $k$ раз:

In [8]:
k = sym.Symbol('k')
fxk = sym.Matrix([[k, 0], [0, k]])
fxk

Matrix([
[k, 0],
[0, k]])

Проекция на $OX$:

In [9]:
sym.Matrix([[1, 0], [0, 0]])

Matrix([
[1, 0],
[0, 0]])

Проекция на $OY$:

In [10]:
sym.Matrix([[0, 0], [0, 1]])

Matrix([
[0, 0],
[0, 1]])

Зеркальная симметрия относительно $OX$:

In [11]:
# знак ординаты меняется на противоположный
sym.Matrix([[1, 0], [0, -1]])

Matrix([
[1,  0],
[0, -1]])

Зеркальная симметрия относительно $OY$:

In [12]:
# знак абсциссы меняется на противоположный
sym.Matrix([[-1, 0], [0, 1]])

Matrix([
[-1, 0],
[ 0, 1]])

Центральная симметрия с центром в $(0,0)$:

In [13]:
sym.Matrix([[-1, 0], [0, -1]])

Matrix([
[-1,  0],
[ 0, -1]])

Поворот на $90^{\circ}$ против часовой:

In [14]:
# 1. поворот против часовой - значит, зеркальная симметрия относительно OY;
# 2. абсцисса и ордината меняются местами
sym.Matrix([[0, -1], [1, 0]])

Matrix([
[0, -1],
[1,  0]])

Поворот на $90^{\circ}$ по часовой:

In [15]:
# 1. поворот по часовой - значит, зеркальная симметрия относительно OX;
# 2. абсцисса и ордината меняются местами
sym.Matrix([[0, 1], [-1, 0]])

Matrix([
[ 0, 1],
[-1, 0]])

***
## <a id='toc1_2_'></a>[__Умножение матриц__](#toc0_)

In [16]:
arr = np.arange(1,16)
X = arr.reshape(3,5)
Y = arr.reshape(5,3)

In [17]:
X

array([[ 1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10],
       [11, 12, 13, 14, 15]])

In [18]:
Y

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12],
       [13, 14, 15]])

In [19]:
X @ Y

array([[135, 150, 165],
       [310, 350, 390],
       [485, 550, 615]])

In [20]:
X[0] @ Y[:,0]

135

In [21]:
X[1] @ Y[:,0]

310

In [22]:
X[0] @ Y[:,1]

150

__`310`__ встает на место $[1,0]$, __`150`__ встает на место $[0,1]$ и т.д.: на $i$-ом месте в $j$-ом столбце матрицы $XY$ стоит произведение $i$-ой строки $X$ и $j$-ого столбца $Y$.

Умножение $XY$ определено, если число столбцов $X$ равно числу строк $Y$. 

***
## <a id='toc1_3_'></a>[__Базис, размерность, ранг__](#toc0_)
__Задача__. Найти координаты вектора $(-4,8,-2)$ в базисе $(1,2,0), (2,0,1), (0,0,2)$.

In [23]:
a = np.array([[1, 2, 0], [2, 0, 0], [0, 1, 2]])
b = np.array([-4, 8, -2])

x = np.linalg.solve(a, b)
x

array([ 4., -4.,  1.])

In [24]:
np.all(a @ x == b)

True

__Задача__. Применить метод Гаусса к заданной матрице __`M`__.

In [25]:
M = sym.Matrix([[0, 6, 8, 17],
                [3, 13, 18, 8],
                [2, 8, 10, 4],
                [1, 3, 2, 0]])

M.rref(pivots=False)

Matrix([
[1, 0, 0, -19/2],
[0, 1, 0,   7/2],
[0, 0, 1,  -1/2],
[0, 0, 0,     0]])

См. [__row echelon form__](https://en.wikipedia.org/wiki/Row_echelon_form).

In [26]:
M.rank()  # 3, что также видно из ref выше

3

***
## <a id='toc1_4_'></a>[__Определитель и его свойства__](#toc0_)
Геометрическая интерпретация: площадь параллелограмма (объем параллелепипеда), натянутого на вектор-__столбцы__ матрицы.

Если столбцы матрицы задают правый базис, ориентированная площадь параллелограмма положительна. Правый базис: кратчайший поворот __против__ часовой стрелки (поворот от большого пальца к указательному на правой руке). В трехмерном случае правая тройка векторов соответствует движению крышки __вверх__ при при ее повороте против часовой стрелки.

__Свойства определителя__:

1. _Свойство равноправности строк и столбцов_. При транспонировании величина определителя сохраняется.

2. _Свойство антисимметрии при перестановке двух строк (или двух столбцов)_. При перестановке двух строк (или двух столбцов) определитель сохраняет свою абсолютную величину, но меняет знак на противоположный.

3. _Линейное свойство определителя_. Если в определителе $n$-ого порядка $\Delta$ некоторая строка является линейной комбинацией двух строк с коэффициентами $\lambda$ и $\mu$, то $\Delta = \lambda \Delta_1 + \mu \Delta_2$. Справедливо для нескольких строк, справедливо для столбцов.

__Следствия из свойств определителя__:

1. _Определитель с двумя одинаковыми строками (или столбцами) равен нулю_. 
<br>При перестановке двух одинаковых строк, с одной стороны, определитель $\Delta$ не изменится, а с другой стороны, изменит знак на противоположный: $\Delta = -\Delta, 2\Delta = 0, \Delta = 0$.

2. _Умножение всех элементов некоторой строки (столбца) определителя на число $\lambda$ равносильно умножению определителя на это число_ $\lambda$. 
<br>Иными словами, общий множитель всех элементов некоторой строки (столбца) определителя можно вынести за знак этого определителя. 
<br>Это вытекает из свойства линейности определителя при $\mu = 0$.

3. _Если все элементы некоторой строки (столбца) определителя равны нулю, то и сам определитель равен нулю_. <br>Это вытекает из предыдущего следствия при $\lambda = 0$.

4. _Если элементы двух строк (столбцов) определителя пропорциональны, то определитель равен нулю_. 
<br>Множитель пропорциональности можно вынести за знак определителя, после чего останется определитель с двумя одинаковыми строками, который равен нулю.

5. _Если к элементам некоторой строки (столбца) определителя прибавить соответствующие элементы другой строки (столбца), умноженные на произвольный множитель $\lambda$, то величина определителя не изменится_. <br>Полученный таким образом определитель можно, согласно свойству линейности определителя, разбить на сумму двух определителей, первый из которых совпадает с исходным, а второй равен нулю, в силу пропорциональности двух строк (столбцов).

***
## <a id='toc1_5_'></a>[__Переход в другой базис__](#toc0_)
### <a id='toc1_5_1_'></a>[__Переход от произвольного базиса к стандартному__](#toc0_)

__Пример__. Пусть $(-1,-2,3)$ это координаты вектора $\vec{u}$ в базисе $\vec{h}_1 = (1,0,2),\vec{h}_2 = (-3,4,0),\vec{h}_3 = (0,-5,8)$. Найти координаты вектора $\vec{u}$ в стандартном базисе.

In [27]:
h = np.array([[1, -3, 0],
              [0, 4, -5],
              [2, 0, 8]])

u = np.array([-1, -2, 3])

h @ u

array([  5, -23,  22])

Чуть более наглядно:

In [28]:
e1, e2, e3 = sym.symbols('e1 e2 e3')  # станд. базис
h1 = e1 + 0*e2 + 2*e3
h2 = -3*e1 + 4*e2 + 0*e3
h3 = 0*e1 - 5*e2 + 8*e3

h = sym.Matrix([h1, h2, h3])
u = sym.Matrix([-1, -2, 3])

h.T @ u

Matrix([[5*e1 - 23*e2 + 22*e3]])

### <a id='toc1_5_2_'></a>[__Переход от стандартного базиса к произвольному__](#toc0_)

__Пример__. Для вектора с координатами $(1,2,3)$ в стандартном базисе найти его координаты в базисе $\vec{f}_1 = (7,0,0),\vec{f}_2 = (6,5,0),\vec{f}_3 = (4,3,2)$.

$Ax = b
\\ A^{-1}Ax = A^{-1}b
\\ x = A^{-1}b$

In [29]:
A = np.array([[7, 6, 4], [0, 5, 3], [0, 0, 2]])
b = np.array([1, 2, 3])

np.linalg.solve(A, b)

array([-0.28571429, -0.5       ,  1.5       ])

In [30]:
x = np.linalg.inv(A) @ b
x

array([-0.28571429, -0.5       ,  1.5       ])

In [31]:
Fraction(x[0]).limit_denominator()

Fraction(-2, 7)

In [32]:
np.all(A @ np.array([-2/7, -1/2, 3/2]) == b)

True

In [33]:
A = sym.Matrix([[7, 6, 4], [0, 5, 3], [0, 0, 2]])
b = sym.Matrix([1, 2, 3])

A.inv() @ b

Matrix([
[-2/7],
[-1/2],
[ 3/2]])

__Следствие__. Если $X$ - квадратная матрица полного ранга, то $\det X^{-1} = \frac{1}{\det X}$:

In [34]:
X = np.array([[4, 7, 0],
              [0, 5, 1],
              [0, 0, 2]])

np.linalg.det(X)

40.0

In [35]:
np.linalg.det(np.linalg.inv(X))

0.025

### <a id='toc1_5_3_'></a>[__Переход от произвольного базиса к другому произвольному__](#toc0_)

__Пример__. Пусть $f : \vec{f_1} = (3,2), \vec{f_2} = (1,0)$ и $g : \vec{g_1} = (5,0), \vec{g_2} = (-1,2)$ - базисы на плоскости. Найти матрицу перехода от $f$ к $g$.

Соединим два случая выше в один пайплайн: переход от произвольного базиса к стандартному (1) и от стандартного к другому произвольному (2).

* случай (1) очевиден и не должен вызывать трудностей.

* случай (2) перевертыш первого: _на что нужно умножить известное, чтобы получить другое известное?_ Простейшее уравнение с учетом матричных особенностей.

* случай (3) - первые два вместе, стандартный базис как перемычка.

In [36]:
F = np.array([[3, 1], [2, 0]])
G = np.array([[5, -1], [0, 2]])

np.linalg.inv(G) @ F

array([[0.8, 0.2],
       [1. , 0. ]])

***
## <a id='toc1_6_'></a>[__Длина, углы, скалярное произведение__](#toc0_)
__Задача__. Найти длину вектора $(2\vec{a}-3\vec{b})$, где $\vec{a} = (9,6,7)$ и $\vec{b} = (5,4,6)$.

In [37]:
a = np.array([9, 6, 7])
b = np.array([5, 4, 6])
c = 2*a - 3*b

In [38]:
np.sqrt(sum(c**2))

5.0

In [39]:
np.linalg.norm(c)

5.0

In [40]:
euclidean(2*a, 3*b)

5.0

### <a id='toc1_6_1_'></a>[__Теорема косинусов__](#toc0_)

$$\Large a^2 + b^2 - 2ab\cos\alpha = c^2$$

(угол $\alpha$ противолежит стороне $c$)

In [41]:
def cosAlpha_tri(a, b, *, c):
    """cos alpha opposing triangle side c. """
    return (a**2 + b**2 - c**2) / (2*a*b)

In [42]:
cosAlpha_tri(7, 9, c=12)

-0.1111111111111111

$(a)\; \langle\vec{x},\vec{x}\rangle = x_1x_1+...+x_nx_n = x_1^2+...+x_n^2 = \|\vec{x}\|^2$

$(b)\; \langle\vec{x}+\vec{y},\vec{z}\rangle = (x_1+y_1)z_1+...+(x_n+y_n)z_n = \langle\vec{x},\vec{z}\rangle + \langle\vec{y},\vec{z}\rangle$

$(c)\; \|\vec{x}+\vec{y}\|^2 = \overbrace{\langle\vec{x}+\vec{y},\vec{x}+\vec{y}\rangle}^{см.(a)} = \underbrace{\langle\vec{x},\vec{x}\rangle + \langle\vec{x},\vec{y}\rangle + \langle\vec{y},\vec{x}\rangle + \langle\vec{y},\vec{y}\rangle}_{см. (b)} = \|\vec{x}\|^2 + 2\langle\vec{x},\vec{y}\rangle + \|\vec{y}\|^2 \Rightarrow$
$$\langle\vec{x},\vec{y}\rangle = \frac{\|\vec{x}+\vec{y}\|^2 - \|\vec{x}\|^2 - \|\vec{y}\|^2}{2}.$$

Из $(c)$ заменой знака:
$$\|\vec{x}-\vec{y}\|^2 = \|\vec{x}\|^2 - 2\langle\vec{x},\vec{y}\rangle + \|\vec{y}\|^2$$

По теореме косинусов:
$$\|\vec{x}-\vec{y}\|^2 = \|\vec{x}\|^2 + \|\vec{y}\|^2 - 2\cdot\|\vec{x}\|\cdot\|\vec{y}\|\cdot\cos\alpha$$

__Вывод__:
$$\Large \langle\vec{x},\vec{y}\rangle = \|\vec{x}\|\cdot\|\vec{y}\|\cdot\cos\alpha$$

__Следствие__: Если вектора ортогональны, их скалярное произведение равно нулю.



In [43]:
def cosAlpha_vec(a, b):
    """cos alpha between given vectors. """
    return a @ b / np.linalg.norm(a) / np.linalg.norm(b)

In [44]:
cosAlpha_vec(np.array([5, 12]), np.array([-3, 4]))

0.5076923076923077

***
## <a id='toc1_7_'></a>[__Ортогональные преобразования__](#toc0_)
_Ортогональным_ называется преобразование $Q$, сохраняющее длины векторов:
<br>$\|Q\vec{x}\| = \|\vec{x}\|$ для всех $\vec{x} \in \R^n$.

Поворот вокруг $\vec{0}$ на угол $\beta$:

In [45]:
a, b, r = sym.symbols('alpha beta r')

rotate = sym.Matrix([[sym.cos(b), -sym.sin(b)],
                     [sym.sin(b), sym.cos(b)]])

vector = sym.Matrix([r*sym.cos(a), r*sym.sin(a)])

(rotate @ vector).applyfunc(lambda x: x.simplify())

Matrix([
[r*cos(alpha + beta)],
[r*sin(alpha + beta)]])

Вектор __`vector`__ перешел из $\begin{pmatrix}
r\cos\alpha\\
r\sin\alpha
\end{pmatrix}$ в $\begin{pmatrix}
r\cos(\alpha+\beta)\\
r\sin(\alpha+\beta)
\end{pmatrix}$ - разумеется, с сохранением длины. Следовательно, матрица поворота __`rotate`__ - ортогональная.

Ортогональное преобразование сохраняет длины и скалярное произведение. Значит, сохраняет и углы. 
<br>Про ортогональное преобразование можно думать как про движение пространства _целиком_: разные части пространства двигаются все вместе и не смещаются друг относительно друга. Поскольку преобразование должно быть линейным, оно оставляет вектор $\vec{0}$ на месте.

В $\R^1\; \text{E}$ или $(-\text{E})$, в $\R^2$ - поворот с центром в $\vec{0}$.

Важнейшее свойство ортогональных матриц:
$$\Large Q^TQ = \text{E},$$
что означает, что $Q^T = Q^{-1}$, и, следовательно, $QQ^T = \text{E}$.
$$1 = \det\text{E} = \det(Q^TQ) = \det Q^T \cdot \det Q = \det Q \cdot \det Q = (\det Q)^2,$$
из чего следует, что определитель любой ортогональной матрицы равен $1$ или $-1$.

***
## <a id='toc1_8_'></a>[__Комплексные числа__](#toc0_)
Число $i$ – не единственное число, которое в квадрате даёт $−1$. Число $(−i)$ в квадрате также дает $(−1)$:
$$(-i)^2 = ((-1) \cdot i)^2 = (-1)^2 \cdot i^2 = 1 \cdot (-1) = -1.$$

$z = a + bi\\
a = \textbf{Re}(z),\,b = \textbf{Im}(z)\\
r = |z| = \sqrt{a^2+b^2}\\
a = r\cos\alpha,\,b = r\sin\alpha\\
\alpha = \arg(z),\,\alpha \in [0,2\pi)$

In [46]:
z = (2 + 3j) * (4 - 5j) + 6 + 7j
z

(29+9j)

In [47]:
z.real, z.imag  # действительная и мнимая части

(29.0, 9.0)

In [48]:
abs(z)  # модуль, r, также np.absolute(z)

30.364452901377952

In [49]:
np.angle(z)  # arg, опции rad/deg, также np.arctan2(z.imag, z.real)

0.3009202343604251

In [50]:
z.conjugate()  # сопряженное компл. число

(29-9j)

На комплексной плоскости сопряжённые числа представлены точками, симметричными относительно действительной оси. Сопряжёнными числами являются корни квадратного уравнения с действительными коэффициентами и отрицательным дискриминантом.

__Свойство__. При умножении комплексных чисел модули перемножаются, а аргументы складываются:

$z = a + bi,\,w = c + di\\
|z| = R,\,|w| = r\\
\arg(z) = \beta,\,\arg(w) = \alpha\\
z \cdot w = (ac - bd) + (bc + ad) \cdot i$

In [51]:
r, R, alpha, beta = sym.symbols('r R alpha beta')
a, b = R * sym.cos(beta), R * sym.sin(beta)
c, d = r * sym.cos(alpha), r * sym.sin(alpha)

In [52]:
(a*c - b*d).simplify()

R*r*cos(alpha + beta)

In [53]:
(b*c + a*d).simplify()

R*r*sin(alpha + beta)

__Следствие__:
$$\Large z^n = R^n(\cos n\beta + i \cdot \sin n\beta)$$

__Задача__. Пусть $z = a + bi$. Найти матрицу отображения, соответствующую $z$.

Напомним вид матрицы поворота вокруг $\vec{0}$ на угол $\beta$:

In [54]:
rotate = sym.Matrix([[sym.cos(beta), -sym.sin(beta)],
                     [sym.sin(beta), sym.cos(beta)]])

Если матрицу __`rotate`__ умножить на $r = |z|$, то ее действие будет эквивалентно умножению на комплексное число $z$: поворот на угол $\beta = \arg(z)$ и растяжение в $r$ раз. Таким образом, искомая матрица отображения: 

In [55]:
a, b = sym.symbols('a b')
by_def = {r * sym.cos(beta): a, r * sym.sin(beta): b}
(r * rotate).subs(by_def)

Matrix([
[a, -b],
[b,  a]])

__Задача__. Найти $(\sqrt{3} + i)^{24}$.

$z = \sqrt{3} + i,\,|z| = 2,\,\arg(z) = \frac{\pi}{6}$:

In [56]:
z = sym.sqrt(3) + I

In [57]:
sym.Abs(z)  # или built-in abs(z)

2

In [58]:
sym.arg(z)

pi/6

Значит, у $(\sqrt{3} + i)^{24}$ модуль равен $2^{24}$, а аргумент равен $24 \cdot \frac{\pi}{6} = 4\pi = 0$. Вывод: $(\sqrt{3} + i)^{24} = 2^{24}$.

Проверка:

In [59]:
factorint((z**24).simplify())

{2: 24}

__Задача__. Найти число, обратное к $(1 + i)$.

$(a + bi)(1 + i) = 1\\
a + ai + bi - b = 1\\
\begin{cases} a + b = 0\\
a - b = 1 \end{cases}\\
2a = 1\\
a = \frac{1}{2},\,b = -\frac{1}{2}$

Искомое число: $\frac{1}{2} - \frac{i}{2}$

Проверка:

In [60]:
(1 + 1j)**(-1)

(0.5-0.5j)

In [61]:
((1 + I)**(-1)).as_real_imag()

(1/2, -1/2)

Модуль числа, обратного числу $z$, равен $\frac{1}{|z|}$, а фаза $(-\arg(z))$: 

In [62]:
def complexInv(z):
    a = -sym.arg(z)
    return (1/sym.Abs(z) * (sym.cos(a) + sym.sin(a) * I)).simplify()

In [63]:
complexInv(1 + I)

1/2 - I/2

__Геометрический смысл__. Если у некоторого числа $z$ модуль $r = |z| > 1$ и фаза $\varphi = \arg(z) > 0$, то:
* умножение на число $z$ __растягивает__ вектор в $r$ раз и поворачивает его на угол $\varphi$ __против__ часовой стрелки;
* умножение на число, сопряженное числу $z$ __`z.conjugate()`__, __растягивает__ вектор в $r$ раз и поворачивает его на угол $\varphi$ __по__ часовой стрелке;
* умножение на число, обратное числу $z$, __сжимает__ вектор в $r$ раз и поворачивает его на угол $\varphi$ __по__ часовой стрелке.

__Основная теорема алгебры__. Любой многочлен с комплексными коэффициентами имеет корень в $\mathbb{C}$.
<br>На математическом языке это свойство называют алгебраической замкнутостью $\mathbb{C}$.

__Следствие__. Любой многочлен степени $n$ с комплексными коэффициентами имеет ровно $n$ корней в $\mathbb{C}$.